src/utils.cpp

```diff
@@ -334,6 +335,621 @@ void cnf_to_dimacs(Term cnf, std::ostringstream & y)
+      int pt = 1;
+      while (true)
+      {
+        try
+        {
+          return solver_->make_symbol("tseitin_to_cnf_" + std::to_string(pt),
+                                      boolsort);
+        }
+        catch (IncorrectUsageException & e)
+        {
+          pt++;
+        }
+      }
+
+    };
+    if (!preorder_)  // using post order traversal as I need the new names of
+                     // the children to generate the new term
+    {
+      smt::Op op = term->get_op();
+      if (!op.is_null())
+      {
+        std::vector<Term> vec;  // a vector of all children
+        for (auto u : term)
+        {
+          Term term_name;
+          bool present = query_cache(
+              u,
+              term_name);  // finding the new name of each child from the cache
+          assert(present == true);
+          vec.push_back(term_name);
+        }
+        Term term_name;
+        query_cache(term, term_name);
+
+        term_name = give_symbolic_name(term);  // making a new symbol
+        save_in_cache(term, term_name);
+
+        reduced.push_back({ term_name, solver_->make_term(op, vec) });
+      }
+      else
+      {  // mapping a symbolic constant to itself
+        save_in_cache(term, term);
+      }
+    }
+
+    return Walker_Continue;
+  }
+};
+
+// binarising xor with multiple children
+class XorBinarize : public IdentityWalker
+{
+ public:
+  XorBinarize(SmtSolver solver_) : IdentityWalker{ solver_, false } {}
+  WalkerStepResult visit_term(Term & term)
+  {
+    if (!preorder_)
+    {
+      smt::Op op = term->get_op();
+      if (!op.is_null())
+      {
+        if (op == smt::Xor)
+        {
+          auto it = term->begin();
+          Term term_name;
+          query_cache((*it),
+                      term_name);  // finding the mapped term from the cache
+          Term ne = term_name;
+          it++;
+          while (it != term->end())
+          {  // Binarising the term
+            Term term_name;
+            query_cache((*it), term_name);
+            ne = solver_->make_term(op, ne, term_name);
+            it++;
+          }
+          save_in_cache(term, ne);  // storing the new binarised term in cache
+        }
+        else
+        {
+          save_in_cache(term, term);
+        }
+      }
+      else
+      {
+        save_in_cache(term, term);
+      }
+    }
+
+    return Walker_Continue;
+  }
+  Term acc_cache(Term term)
+  {
+    Term ne;
+    query_cache(term, ne);
+    return ne;
+  }
+};
+
+// Given a boolean formula, removes terms "true" and "false" to give a formula
+// without adding new symbols
+class EliminateBooleanConstants : public IdentityWalker
+{
+ public:
+  EliminateBooleanConstants(SmtSolver solver_)
+      : IdentityWalker{ solver_, false }
+  {
+  }
+  WalkerStepResult visit_term(Term & term)
+  {
+    auto is_true = [&](Term t) {  // If the term is "true"
+      return ((t->to_string()) == "true");
```

Comment:
It is better not to use string comparisons but term comparisons. You can create beforehand a `true` term and a `false` term, and compare to them. Maybe some day we decide to print this as `TRUE` or something like that, and then this code will break. The value of `t` should not depend on its string representation.